### PR TITLE
Implement forums

### PR DESCRIPTION
**Pull request by HopeBaron**
relevant doc PRs:
- [ ] https://github.com/discord/discord-api-docs/pull/4761
  - [x] `ChannelType.GuildForum`
  - [x] `ChannelFlag` + `ChannelFlags`
  - [x] `DiscordChannel.flags` and `ChannelData.flags`
  - [ ] core channel type that is a `ThreadParentChannel` (keep `activeThreads` and `getPublicArchivedThreads()`) but no `TopGuildMessageChannel`, however it also needs a new thread creation method, as it's different from both creating a thread from an existing message and without a message: threads created in forum channels always also create a message -> new builder that is a message create builder and a thread create builder, see https://discord.com/developers/docs/resources/channel#start-thread-in-forum-channel
  - [x] errors 10087, 30047, 30048, 40061
  - [x] track `ThreadCreate` events to update cached `ChannelData.lastMessageId` (same as `MessageCreate` for normal message channels)
- [ ] https://github.com/discord/discord-api-docs/pull/4850
  - [ ] creating threads in forums returns thread and message
- [x] https://github.com/discord/discord-api-docs/pull/4867
       (only the new [`flags`](https://github.com/advaith1/discord-api-docs/blob/3999cf9e7c4e68ba97907b3af1986066b4e9a3fb/docs/resources/Channel.md?plain=1#L836) fields in `docs/resources/Channel.md` is relevant from this PR)
  - [x] allow changing flags in thread modify builder (for pinning threads in forums)
- [x] https://github.com/discord/discord-api-docs/pull/5007
  - [x] `WebhookExecuteRequest.threadName` and `WebhookMessageCreateBuilder.threadName`
- [x] https://github.com/discord/discord-api-docs/pull/5018
- [ ] https://github.com/discord/discord-api-docs/pull/5206
  - [ ] `DiscordChannel.totalMessageSent` and `ChannelData.totalMessageSent` -> also add on core thread channel types
  - [ ] `DiscordMessge.position`, `MessageData.position`, `Message.position`
  - [ ] update cached values for `messageCount` and `totalMessageSent` on `MessageCreate`, `MessageDelete` and `MessageDeleteBulk`
- [ ] https://github.com/discord/discord-api-docs/pull/5365
       (not sure if this one is relevant)
- [ ] https://github.com/discord/discord-api-docs/pull/5391
  - [ ] `availableTags`, `appliedTags`, `defaultReactionEmoji` and `defaultThreadRateLimitPerUser` for `DiscordChannel`, `ChannelData` and on appropriate core channel types
  - [x] `availableTags`, `defaultReactionEmoji` and `defaultThreadRateLimitPerUser` for channel modify builders
  - [x] `ChannelFlag.RequireTag`
  - [x] "Default Reaction" and "Forum Tag" types
  - [x] `appliedTags` in builders for starting and modifying threads in forum channels
  - [x] errors 40058, 40066, 40067, 220001, 20002, 20004
- [x] https://github.com/discord/discord-api-docs/pull/5474
  - [x] `availableTags`, `defaultReactionEmoji` and `defaultThreadRateLimitPerUser` for channel create builders
- [x] https://github.com/discord/discord-api-docs/pull/5469
  - [x] allow changing flags in channel modify builder (for requiring tags in forums)
- [ ] https://github.com/discord/discord-api-docs/pull/5488
  - [x] `SortOrderType` kord enum
  - [x] `DiscordChannel.defaultSortOrder` and `ChannelData.defaultSortOrder` -> also add on core forum channel type
  - [x] `defaultSortOrder` in forum create / modify builders
- [ ] https://github.com/discord/discord-api-docs/pull/5458
- [ ] https://github.com/discord/discord-api-docs/pull/5478
- [x] https://github.com/discord/discord-api-docs/pull/5745
  - [x] `ForumLayoutType` kord enum
  - [x] `DiscordChannel.defaultForumLayout` and `ChannelData.defaultForumLayout` -> also add on core forum channel type
  - [x] `defaultForumLayout` in builders

### CODE REVIEW

**Review by Lukellmann — CHANGES_REQUESTED** [review]
First thoughts, haven't looked at everything yet.

File: common/src/main/kotlin/entity/DiscordChannel.kt [hunk 1]
```diff
@@ -55,6 +55,8 @@ public data class DiscordChannel(
     val name: Optional<String?> = Optional.Missing(),
     val topic: Optional<String?> = Optional.Missing(),
     val nsfw: OptionalBoolean = OptionalBoolean.Missing,
+    @SerialName("nsfw_level")
+    val nsfwLevel: NsfwLevel,
```
[the remark this answers is not in the document]
Comment (HopeBaron) [reply common/src/main/kotlin/entity/DiscordChannel.kt:59]: seems like i misunderstood this https://discord.com/channels/556525343595298817/1014507143988265081

File: common/src/main/kotlin/entity/DiscordChannel.kt [hunk 2]
```diff
@@ -159,6 +161,9 @@ public sealed class ChannelType(public val value: Int) {
      */
     public object GuildDirectory : ChannelType(14)
 
+    public object GuildForum : ChannelType(15)
```
[the remark this answers is not in the document]
Comment (HopeBaron) [reply common/src/main/kotlin/entity/DiscordChannel.kt:164]: I wonder if we can even merge these togather

[the remark this answers is not in the document]
Comment (Lukellmann) [reply common/src/main/kotlin/entity/DiscordChannel.kt:164]: why not? 🤔

File: common/src/main/kotlin/entity/DiscordChannel.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (Lukellmann) [reply common/src/main/kotlin/entity/DiscordChannel.kt]: I removed it again

File: common/src/main/kotlin/entity/DiscordChannel.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (Lukellmann) [reply common/src/main/kotlin/entity/DiscordChannel.kt]: Did the merge

File: core/src/main/kotlin/behavior/channel/ForumChannelBehavior.kt [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+import dev.kord.core.behavior.channel.threads.ThreadParentChannelBehavior
+import dev.kord.core.cache.data.ChannelData
+import dev.kord.core.entity.channel.Channel
+import dev.kord.core.entity.channel.ForumChannel
+import dev.kord.rest.builder.channel.ForumChannelModifyBuilder
+import dev.kord.rest.service.patchForumChannel
+import kotlin.contracts.InvocationKind
+import kotlin.contracts.contract
+
+public interface ForumChannelBehavior : ThreadParentChannelBehavior {
```
[the remark this answers is not in the document]
Comment (dookon) [reply core/src/main/kotlin/behavior/channel/ForumChannelBehavior.kt:12]: Applied at https://github.com/kordlib/kord/pull/709

File: core/src/main/kotlin/entity/channel/Channel.kt [hunk 4]
```diff
@@ -26,6 +27,9 @@ public interface Channel : ChannelBehavior {
      */
     public val type: ChannelType get() = data.type
 
+    /** The flags of this channel, if present. */
+    public val flags: ChannelFlags? get() = data.flags.value
```
Comment (dookon) [inline core/src/main/kotlin/entity/channel/Channel.kt:31]: So far, only forum threads have flags, wouldn't it be better to move it to his entity?

Comment (Lukellmann) [reply core/src/main/kotlin/entity/channel/Channel.kt:31]: I think for flags it's ok to be located in the base `Channel` type. They might add new `ChannelFlag`s in the future...

**Pull request merged**